src/stream/src/executor/watermark_filter.rs

```diff
@@ -0,0 +1,477 @@
+
+                    // Empty chunk should not be processed.
+                    if chunk.cardinality() == 0 {
+                        continue;
+                    }
+
+                    let watermark_array = watermark_expr
+                        .eval_infallible(chunk.data_chunk(), |err| {
+                            ctx.on_compute_error(err, &info.identity)
+                        });
+
+                    // Build the expression to calculate watermark filter.
+                    let watermark_filter_expr = Self::build_watermark_filter_expr(
+                        watermark_type.clone(),
+                        event_time_col_idx,
+                        current_watermark.clone(),
+                    )?;
+
+                    // Assign a new watermark.
+                    for watermark in watermark_array.iter() {
+                        // Ignore the Null watermark
+                        if watermark.is_some() && to_datum_ref(&current_watermark) < watermark {
+                            current_watermark = watermark.to_owned_datum();
+                        }
+                    }
+
+                    let pred_output = watermark_filter_expr
+                        .eval_infallible(chunk.data_chunk(), |err| {
+                            ctx.on_compute_error(err, &info.identity)
+                        });
+
+                    if let Some(output_chunk) = SimpleFilterExecutor::filter(chunk, pred_output)? {
+                        yield Message::Chunk(output_chunk);
+                    };
+
+                    yield Message::Watermark(Watermark::new(
+                        event_time_col_idx,
+                        current_watermark.clone(),
+                    ));
+                }
+                Message::Watermark(watermark) => {
+                    if watermark.col_idx != event_time_col_idx {
+                        yield Message::Watermark(watermark)
+                    } else {
+                        bail!("WatermarkFilterExecutor should not receive a watermark on the event it is filtering.")
+                    }
+                }
+                Message::Barrier(barrier) => {
+                    // Update the vnode bitmap for state tables of all agg calls if asked.
+                    if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(ctx.id) {
+                        let previous_vnode_bitmap = table.update_vnode_bitmap(vnode_bitmap.clone());
+
+                        // Take the global max watermark when scaling happens.
+                        if previous_vnode_bitmap != vnode_bitmap {
+                            current_watermark =
+                                Self::get_global_max_watermark(&table, watermark_type.clone())
+                                    .await?;
+                        }
+                    }
+
+                    if barrier.checkpoint {
+                        // Persist the watermark when checkpoint arrives.
```

Comment:
We can check whether `current_watermark` == `last_checkpoint_watermark` here. If equal, we can skip here.

Reply:
Sure. Although it is unlike like to hit on the optimization in most cases, and the cost of writing 256 int values is fairly low.